geps/gep-91/index.md

```diff
@@ -1,22 +1,135 @@
 
 ## Non-Goals
-- Define other fields that can be used to verify the client certificate such as the Certificate Hash or Subject Alt Name. 
+- Define other fields that can be used to verify the client certificate such as the Certificate Hash. 
+
+## Existing support in Implementations
+
+This feature is widely supported in implementations that support Gateway API.
+This table highlights the support. Please feel free to add any missing implementations not mentioned below.
+
+| Implementation | Support       |
+|----------------|------------|
+| Apache APISIX  | [ApisixTls.Client.CASecret](https://apisix.apache.org/docs/ingress-controller/tutorials/mtls/#mutual-authentication)      |
+| Contour        | [HTTPProxy.Spec.VirtualHost.Tls.ClientValidation.CASecret](https://projectcontour.io/docs/v1.17.1/config/tls-termination/)      |
+| Emissary Ingress| [TlSContext.Spec.Secret](https://www.getambassador.io/docs/emissary/latest/topics/running/tls/mtls)     |
+| Gloo Edge      | [VirtualService.Spec.SSLConfig.SecretRef](https://docs.solo.io/gloo-edge/latest/guides/security/tls/server_tls/#configuring-downstream-mtls-in-a-virtual-service)      |
+| Istio          | [Gateway.Spec.Servers.TLS.Mode](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway)      |
+| Kong           | [mTLS Plugin](https://docs.konghq.com/hub/kong-inc/mtls-auth/)      |
+| Traefik        | [TLSOption.Spec.ClientAuth](https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls)    |
+
+### API
+
+* Introduce a `Client` field of type `TLSClientContext` within [GatewayTLSConfig][] to hold TLS configuration specific to the client.
+* Introduce a `Validation` field of type `TLSValidationContext` within the `Client` field that can be used to validate the peer with which the TLS connection is being made.
+to the Gateway.
+* Introduce a `caCertificateRefs` field within `ValidationContext` that can be used to specify a list of CA Certificates that
+can be used as a trust anchor to validate the certificates presented by the client.
+* This new field is mutually exclusive with the [BackendTLSPolicy][] configuation which is used to validate the TLS certificate presented by the peer on the connection between the Gateway and the backend, and this GEP is adding support for validating the TLS certificate presented by the peer on the connection between the Gateway and the downstream client.
+* Introduce an optional `subjectAltNames` field within `ClientValidationContext` that can be used to specify one or more alternate names to verify the subject identity in the certificate presented by the client. The maximum number of alternate names that can be specified is implementation defined.
```

Comment:
I am not sure about this one - this is really authorization policy, and having a list of identities in the gateway itself is perhaps the worse for a gateway ( scale, personas doing the config, etc).

Reply:
In Istio that's part of Authorization policy.

Reply:
@frankbu and @howardjohn from Istio requested for adding subjectAltName

Reply:
Well, Istio does have this field in istio Gateway - but we also have the authorization policy as a separate resource. 

I don't think this is the best UX for verifying client certificates - do we have any real world use case where client certificates are using SANs ( real internet clients - not mesh ), and real internet gateways authenticate by listing all the clients in the config file ? Even the old httpasswd was a separate config and didn't scale well...

Reply:
@arkodg Without a solid use case, it's best to drop `subjectAltNames` from the API to move this PR forward.

Reply:
@howardjohn or @frankbu can you provide more info on why you'd like SANs to be configurable at this level?

Reply:
Good question about real world use cases. I was only requesting this feature because Istio supports it and we'd rather not add an Istio-specific gateway-api extension if we can avoid it. https://github.com/istio/istio/pull/48013.

@howardjohn?

Reply:
I generally agree this probably makes sense to split out as a more general authz. However, there is no authz in gateway-api so its a bit tricky..

Generally any TLS usage ought to actually verify the identity in the certificates; rarely is it sufficient to just validate the issuing CA. So we would end up in a state where to properly use this feature, each vendor would need to have its own policy and users would need to use it, or be left in an insecure state.

Reply:
removing SAN for now, will add a comment in the GEP mentioning the reason - "it falls under the authz bucket so we can revisit this field once the project tackles authz as a whole"

Reply:
Knative uses these features for the gateway=>backend. Since those requests can be handled by a revision instance or the activator component (when we scale up).

We don't want our activator component impersonating all revision instances and having to deal with so many certificates)

Reply:
On the client=>gateway side wouldn't allowing subjectAltNames enable auth with certs?

eg. My internal corp site requires my machine's client cert to match the alternate names `*.internal.broadcom.net` or `*.internal.vmware.com` ?